src/test/scala/chisel3/internal/NameCollisionSpec.scala

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chiselTests
+
+import chisel3._
+import chisel3.stage.ChiselStage
+
+class NameCollisionSpec extends ChiselFlatSpec with Utils {
+  behavior.of("Builder")
+
+  it should "error on duplicated names with a correct message" in {
+    (the[ChiselException] thrownBy extractCause[ChiselException] {
+      (new ChiselStage).emitChirrtl(
+        new Module {
+          val foo, bar = IO(Input(UInt(8.W)))
+          val out      = IO(Output(UInt(8.W)))
+
+          // Rename both inputs to the same name
+          foo.suggestName("same")
+          bar.suggestName("same")
+
+          out := foo + bar
+        },
+        
+        Array("--throw-on-first-error")
+      )
+    }).getMessage should include("Attempted to name NameCollisionSpec_Anon.same: IO[UInt<8>] with a duplicated name 'same'.")
+  }
+
+  it should "error on sanitization resulting in duplicated names with a helpful message" in {
+    // Case one: An unsanitary name that results in a collision with an existing name once sanitized
+    (the[ChiselException] thrownBy extractCause[ChiselException] {
+      (new ChiselStage).emitChirrtl(
+        new Module {
+          val foo, bar = IO(Input(UInt(8.W)))
+          val out      = IO(Output(UInt(8.W)))
+
+          // Seed an initial name with no necessary sanitization
+          foo.suggestName("unsanitary")
+          // Seed an additional name which, when sanitized, collides with the previous name
+          bar.suggestName("unsanitary-")
+
+          out := foo + bar
+        },
+        
+        Array("--throw-on-first-error")
+      )
+    }).getMessage should include("Attempted to name NameCollisionSpec_Anon.unsanitary-: IO[UInt<8>] with an unsanitary name 'unsanitary-'")
```

Comment:
does this need a collision? should we have a case without the collision where it still yells at you for the unsanitary name?

Reply:
oh wait that was the next one.

So shouldn't this error message not just tell you had an unsanitary name, but "after sanitizing, it collides with `unsanitary`?

it would be nice if you got a source locator for the port you were using and the port you are colliding with...

Reply:
ok i looked at the full error message (i was just reading the tests) and you cover this

Reply:
> it would be nice if you got a source locator for the port you were using and the port you are colliding with...

I agree but that requires more extensive internals work, this PR at least makes the error messages better in the meantime.